crates/runtime/src/databackend/viewtable.rs

```diff
@@ -0,0 +1,98 @@
+use std::{pin::Pin, sync::Arc, thread, time::Duration};
+
+use datafusion::{
+    datasource::ViewTable,
+    error::DataFusionError,
+    execution::context::SessionContext,
+    sql::{parser::DFParser, sqlparser::dialect::AnsiDialect},
+};
+use futures_core::Future;
+use snafu::prelude::*;
+
+use super::{
+    DataBackend, DataBackendType, Result, UnableToAddDataSnafu, UnableToCreateTableDataFusionSnafu,
+    UnableToCreateTableSnafu, UnableToParseSqlSnafu, UnsupportedOperationSnafu,
+};
+
+pub struct ViewTableBackend {
+    _ctx: Arc<SessionContext>,
+    _name: String,
+    _sql: String,
+}
+
+impl ViewTableBackend {
+    pub async fn new(ctx: Arc<SessionContext>, name: &str, sql: &str) -> Result<Self> {
+        let statements = DFParser::parse_sql_with_dialect(sql, &AnsiDialect {})
+            .context(UnableToParseSqlSnafu)?;
+        if statements.len() != 1 {
+            return UnableToCreateTableSnafu {
+                reason: format!(
+                    "Expected 1 statement to create view, received {}",
+                    statements.len()
+                )
+                .to_string(),
+            }
+            .fail();
+        }
+
+        let view: ViewTable;
+        let mut backoff = Duration::from_secs(1);
+        let mut attempts = 1;
+        // Tables are currently lazily created (i.e. not created until first data is received) so that we know the table schema.
+        // This means that we can't create a view on top of a table until the first data is received for all dependent tables and therefore
+        // the tables are created. To handle this, if view creation fails with a plan error, we retry until the table is created.
+        loop {
+            let plan_result = ctx.state().statement_to_plan(statements[0].clone()).await;
+            match plan_result {
+                Ok(plan) => {
+                    view = ViewTable::try_new(plan, Some(sql.to_string()))
+                        .context(UnableToAddDataSnafu)?;
+                    break;
+                }
+                Err(e) => match e {
+                    DataFusionError::Plan(_) => {
+                        if attempts > 5 {
+                            return Err(e).context(UnableToCreateTableDataFusionSnafu);
+                        }
+
+                        tracing::error!(
+                            "Plan error for {}, retrying in {:?}: {}",
+                            name,
+                            backoff,
+                            e
+                        );
+                        thread::sleep(backoff);
```

Comment:
We can be smarter about how we do this - instead of retrying on a timer, we can parse the SQL, extract out the relations that are referenced, and check if those tables are registered but not populated yet. If they are, then we connect a channel that only retries once the source datasets receive data (via a mpsc channel)

Reply:
Yes, see my handoff. However, I think that's a larger piece of work and not necessary to have views working. I don't want to get stuck in the weeds here building out a system for publishing and subscribing to dataset updates. I think it's best to think about that separately.

Reply:
As discussed offline: we are going to parse the SQL to determine the dependent tables and if they exist. If they do exist, then we will listen to some channel that publishes when new data is received for that dataset and try again once new data is received.

Reply:
Spoke further offline and decided we're over-engineerning this problem. The ideal solution is likely building a dependency graph so we know what datasets need to be initialized first and which can be initialized in parallel. For now, by parsing out table names and waiting for those tables to exist we are simulating this behavior.